Applications/Android/NNDetector/app/src/main/jni/simpleshot.cpp

```diff
@@ -0,0 +1,444 @@
+                                     {"batch_size=1", "epochs=1"});
+
+  LayerHandle backbone_layer = ml::train::layer::BackboneTFLite(
+    {"name=backbone", "model_path=" + det_backbone_path,
+     "input_shape=" + in_str + ":" + in_str + ":3", "trainable=false"});
+  det_model->addLayer(backbone_layer);
+
+  det_model->summarize(std::cout, ML_TRAIN_SUMMARY_MODEL);
+
+  std::shared_ptr<ml::train::Optimizer> optimizer;
+  optimizer = ml::train::optimizer::SGD({"learning_rate=0.1"});
+
+  if (det_model->setOptimizer(optimizer)) {
+    throw std::invalid_argument("failed to set optimizer");
+  }
+  if (det_model->compile()) {
+    throw std::invalid_argument("model compilation failed");
+  }
+  if (det_model->initialize()) {
+    throw std::invalid_argument("model initiation failed");
+  }
+
+  return det_model.get();
+}
+
+ml::train::Model *initialize_rec(int argc, char *argv[]) {
+
+  if (argc < 5) {
+    std::cerr << "usage: ./main [rec backbone path] "
+                 "[rec input image dim] [num class] [knn norm variant]\n";
+    throw std::invalid_argument("wrong argument");
+  }
+
+  std::string rec_backbone_path = argv[1];
+  unsigned int rec_input_img_dim = std::stoul(argv[2]);
+  unsigned int num_class = std::stoul(argv[3]);
+  std::string knn_norm_variant = argv[4];
+
+  ANDROID_LOG_D("---recognition backbone path: %s", rec_backbone_path.c_str());
+  ANDROID_LOG_D("---num classes: %d", num_class);
+  ANDROID_LOG_D("---input shape: %d:%d:3", rec_input_img_dim,
+                rec_input_img_dim);
+  ANDROID_LOG_D("---knn norm variant: %s", knn_norm_variant.c_str());
+
+  rec_model = ml::train::createModel(ml::train::ModelType::NEURAL_NET,
+                                     {"batch_size=1", "epochs=1"});
+
+  LayerHandle backbone_layer = ml::train::layer::BackboneTFLite(
+    {"name=backbone", "model_path=" + rec_backbone_path,
+     "input_shape=" + std::to_string(rec_input_img_dim) + ":" +
+       std::to_string(rec_input_img_dim) + ":3",
+     "trainable=false"});
+  rec_model->addLayer(backbone_layer);
+
+  auto generate_knn_part = [num_class](const std::string &variant_) {
+    std::vector<LayerHandle> v;
+
+    const std::string num_class_prop = "num_class=" + std::to_string(num_class);
+
+    if (variant_ == "UN") {
+      /// left empty intended
+    } else if (variant_ == "L2N") {
+      LayerHandle l2 = ml::train::createLayer(
+        "preprocess_l2norm", {"name=l2norm", "trainable=false"});
+      v.push_back(l2);
+    } else {
+      std::stringstream ss;
+      ss << "unsupported variant type: " << variant_;
+      throw std::invalid_argument(ss.str().c_str());
+    }
+
+    LayerHandle knn = ml::train::createLayer(
+      "centroid_knn", {"name=knn", num_class_prop, "trainable=false"});
+    v.push_back(knn);
+
+    return v;
+  };
+
+  auto knn_part = generate_knn_part(knn_norm_variant);
+  for (auto &layer : knn_part) {
+    rec_model->addLayer(layer);
+  }
+  rec_model->summarize(std::cout, ML_TRAIN_SUMMARY_MODEL);
```

Comment:
I think we do not need this call unless we implement this for android log. Currently, it produces std::out

Reply:
done